util/devel/check_paths

```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/env python
+
+# Copyright 2004-2018 Cray Inc.
+# Other additional copyright holders may be indicated within.
+#
+# The entirety of this work is licensed under the Apache License,
+# Version 2.0 (the "License"); you may not use this file except
+# in compliance with the License.
+#
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Try to identify references to non-existent files, e.g. doc/foo
+instead of doc/rst/foo.  This script recognizes potential file
+references by recognizing certain strings as likely to be the prefix
+of a path in the Chapel repository, e.g., "doc/" or "modules/" or
+"test/".
+"""
+
+import os.path
+import argparse
+
+def check_path(p):
+    """returns True if p exists, False if it doesn't"""
+
+    # These are common written-out alternatives, not paths
+    if p == "compiler/runtime":
+        return True
+    if p == "compiler/module":
+        return True
+
+    # These show up in generated html files relative to a "modules"
+    # directory that isn't the top-level one we're going to look in.
+    if p == "modules/chpldoc.doc.html":
+        return True
+    if p == "modules/chpldoc.doc/Defined.html":
+        return True
+    # Or in generated .js
+    if p == "modules/chpldoc.doc" or p == "modules/chpldoc.doc/Defined":
+        return True
+    # jquery .js file means a different test/ directory too
+    if p == "test/unit/core.js":
+        return True
+
+    return os.path.isfile(p) or os.path.isdir(p)
+
+# Memoize will be more useful once the script runs on multiple files
+# in one go.
+# Cribbed from a stackoverflow answer
+class Memoize:
+    def __init__(self, f):
+        self.f = f
+        self.memo = {}
+    def __call__(self, *args):
+        if not args in self.memo:
+            self.memo[args] = self.f(*args)
+        return self.memo[args]
+
+@Memoize
+def check_token(t):
+    """If t "looks like" a path but doesn't exist, return False.
+    Otherwise, return True."""
+
+    # Skip everything that doesn't look like something
```

Comment:
Could we also add an exclusion list? e.g. `doc/html` should probably be skipped, since it blows up with false positives.

Reply:
Oh, I guess I don't see that because I don't have docs built in ``doc/html``.

When the script learns to walk, it can skip that tree.  But right now, it's invoked specifically on a file, and it seems in poor form to just silently skip it.  It would be better to change the invocation to exclude that subdir.  That would make the example invocation more complicated, something involving ``find``'s ``-prune`` command, which it always takes me a long time to figure out.  :(